### PR TITLE
PIM-7143: Mass delete products and product models with job

### PR DESCRIPTION
**Pull request by willy-ahva**
**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | OK
| Added legacy Behats               | OK
| Added acceptance tests            | -
| Added integration tests           | -
| Changelog updated                 | OK
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo !
| Migration script                  | OK
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: CHANGELOG-2.2.md [hunk 1]
```diff
@@ -1,3 +1,13 @@
+# 2.2.x
+
+## Enhancements
+
+- PIM-7143: Be able to delete products and product models in mass using a backend job
+
+## New jobs
+IMPORTANT: In order for your PIM to work properly, you will need to run the following commands to add the missing job instances.
+- Add the job instance `delete_products_and_product_models`: `bin/console akeneo:batch:create-job "Akeneo Mass Edit Connector" "delete_products_and_product_models" "mass_delete" "delete_products_and_product_models" '{}' "Mass delete products" --env=prod`
```
Comment (SamirBoulil) [inline CHANGELOG-2.2.md]: Don't forget the permissions on EE

File: src/Pim/Component/Enrich/Job/DeleteProductsAndProductModelsTasklet.php [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+        }
+
+        $this->stepExecution->addSummaryInfo('deleted_products', 0);
+        $this->stepExecution->addSummaryInfo('deleted_product_models', 0);
+
+        $products = $this->getProductsCursor();
+
+        $loopCount = 0;
+        $entitiesToRemove = [];
+        while ($products->valid()) {
```
Comment (SamirBoulil) [inline src/Pim/Component/Enrich/Job/DeleteProductsAndProductModelsTasklet.php]: You could have used a foreach here no ?

File: src/Pim/Component/Enrich/Job/DeleteProductsAndProductModelsTasklet.php [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+        $products = $this->filterProducts($entities);
+        $productModels = $this->filterProductModels($entities);
+
+        $this->productRemover->removeAll($products);
+        $this->stepExecution->incrementSummaryInfo('deleted_products', count($products));
+
+        $this->productModelRemover->removeAll($productModels);
+        $this->stepExecution->incrementSummaryInfo('deleted_product_models', count($productModels));
+
+        $this->cacheClearer->clear();
```
Comment (SamirBoulil) [inline src/Pim/Component/Enrich/Job/DeleteProductsAndProductModelsTasklet.php]: This is not part of the deleting process, maybe you can do it in the execute (yes twice it's ok for me).

Comment (willy-ahva) [reply src/Pim/Component/Enrich/Job/DeleteProductsAndProductModelsTasklet.php]: I wondered if it was a good solution but I decided to regroup all deletion process (removeAll, summary info and cache clear) in one method.

File: src/Pim/Component/Enrich/Job/DeleteProductsAndProductModelsTasklet.php (hunk 2 above)
Comment (willy-ahva) [reply src/Pim/Component/Enrich/Job/DeleteProductsAndProductModelsTasklet.php]: yes why not, it's ok for me like that

File: src/Pim/Bundle/DataGridBundle/Resources/public/js/datagrid/action/mass-delete-action.js [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+            /**
+             * Converts grid data into pqb filters and gathers job instance code, actions and items count.
+             *
+             * @return {Promise}
+             */
+            getData: function () {
+                let actionParameters = this.getActionParameters();
+                actionParameters.actionName = this.route_parameters['actionName'];
+                actionParameters.gridName = this.route_parameters['gridName'];
+                const query = '?' + $.param(actionParameters);
```
Comment (juliensnz) [inline src/Pim/Bundle/DataGridBundle/Resources/public/js/datagrid/action/mass-delete-action.js]: you can use string template ````?${$.param(actionParameters)}````

File: src/Pim/Bundle/DataGridBundle/Resources/public/js/datagrid/action/mass-delete-action.js [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+                })
+                .then(() => {
+                    router.redirectToRoute(this.config.backRoute);
+
+                    messenger.notify(
+                        'success',
+                        __(
+                            this.config.launchedLabel,
+                            {
+                                operation: 'mass delete'
```
Comment (juliensnz) [inline src/Pim/Bundle/DataGridBundle/Resources/public/js/datagrid/action/mass-delete-action.js]: don't we use a translation key

File: src/Pim/Component/Enrich/Job/DeleteProductsAndProductModelsTasklet.php (hunk 2 above)
Comment (juliensnz) [reply src/Pim/Component/Enrich/Job/DeleteProductsAndProductModelsTasklet.php]: With cursor, we often use while statement as the end of the loop is defined by the valid operator. I'm ok with both. BTW, this is a implementation detail, it's not really important here

File: src/Pim/Bundle/DataGridBundle/Resources/public/js/datagrid/action/mass-delete-action.js [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+                MassAction.prototype.initialize.apply(this, arguments);
+            },
+
+            /**
+             * Displays a confirm dialog and mass delete if action is confirmed.
+             */
+            execute: function() {
+                this.getData().then((data) => {
+                    this.data = data;
```
Comment (juliensnz) [inline src/Pim/Bundle/DataGridBundle/Resources/public/js/datagrid/action/mass-delete-action.js]: You can make this action stateless by passing data to 'getConfirmDialog'

File: src/Pim/Bundle/DataGridBundle/Resources/public/js/datagrid/action/mass-delete-action.js [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+            /**
+             * Get view for confirm modal.
+             *
+             * @return {oro.Modal}
+             */
+            getConfirmDialog: function() {
+                this.confirmModal = Dialog.confirmDelete(
+                    __(this.config.confirmLabel),
+                    __('pim_enrich.confirmation.delete_item'),
+                    this.doMassDelete.bind(this),
```
Comment (juliensnz) [inline src/Pim/Bundle/DataGridBundle/Resources/public/js/datagrid/action/mass-delete-action.js]: then pass data to `doMassDelete`

File: src/Pim/Bundle/DataGridBundle/Resources/public/js/datagrid/action/mass-delete-action.js (hunk 4 above)
Comment (willy-ahva) [reply src/Pim/Bundle/DataGridBundle/Resources/public/js/datagrid/action/mass-delete-action.js]: didn't know this ty

**Pull request merged**